34: replace url crate with peg parser based on RFC 3986

jac18281828 opened the pull request:
closes #34 

- adding path + query + fragment
- test coveragee
- parses query as a vec of string tuple

general cleanup, make DidUrl immutable by removing setter

fix rustdoc

insipx commented:
🧹 `url` 🧹

nice! I like the contextual errors for parsing, I hope it will make it easier on users if they get the DID wrong

37ng reviewed (commented):
Would it make sense to use map instead of an array for queries? It's more efficient to search keys.
Also, would it make sense to test immutability? i.e. checking that original object is not modified after calling `with_[query/path/attribute/...]`?

37ng commented on lib/src/types/did_parser.rs:
````diff
@@ -28,30 +53,94 @@ peg::parser! {
         /// ```
         pub rule ethr_did() -> Did
-            = method:method() ":" network:network()? account:account() { Did { method, network: network.unwrap_or_default(), account } }
+            = did() ":" method:method() ":" network:(chain_id() / network())? account:account() { Did { method, network: network.unwrap_or_default(), account } }
````
In the old version, the network will be defaulted to `mainnet` when eluded. I assume we don't want to keep that logic now? I'm fine with that tho

jac18281828 commented:
> Would it make sense to use map instead of an array for queries? It's more efficient to search keys. Also, would it make sense to test immutability? i.e. checking that original object is not modified after calling `with_[query/path/attribute/...]`?

It's a good idea to use the map but it will not work because the names and values are not distinct in the specification. A query could have a=a, a=a, a=a and be valid syntax its application dependent what the meaning of that is.   Imagine the protocol is 'execute this instruction'.  

The performance of an array vs a map for 2-3 elements is probably actually better for lookup in a typical application where every parameter will have meaning.      Since we are using 1 at the moment we can probably ignore this issue.

The test of immutability should be guaranteed by rust ownership rules, but I will add a check for sanity.

37ng commented:
> The test of immutability should be guaranteed by rust ownership rules, but I will add a check for sanity.

In unit tests, imo it's crucial not to depend on implementation specifics. For functions with_*** functions, they should ideally return a new object rather than altering the original one. Our tests must verify this behavior directly, avoiding assumptions about Rust's clone() method or omitting such tests.

Agreed with the query data structure, thanks for explaining.

jac18281828 commented:
Please take a last look.   I believe it's ready to go.  I updated with some cleanup:

* removed tiny-keccak
* added parsing support with a test for public key did....I suspect pks are still not working and we need to add a ticket to test that and verify it is all good.
* cleaned up account creation
* added coverage for codecov complaints

37ng commented on lib/src/types/did_url.rs:
```diff
@@ -1,20 +1,30 @@
+ * Converts a public key to an ethereum address.  The last 20 bytes
+ * of the keccak256 hash of the public key is returned
+ */
+fn public_key_to_address(key: &[u8]) -> String {
```
Would this [function](https://docs.ethers.org/v5/api/utils/address/#utils-computeAddress) from ethers be helpful?

jac18281828 replied:
The reason I didn't use that is it requires a wallet.    It felt ceremonial to instantiate all that when we don't require the wallet ourselves.

37ng commented on lib/src/types/did_url.rs:
```diff
@@ -54,7 +64,37 @@ impl ToString for Account {
+}
+
+impl Account {
+    pub fn from_hex_key(key: Vec<u8>) -> Self {
```
maybe use From would be more idiomatic?
```rust
impl From<HexKey/Address/&str> for Account {
    // implementation
}
```

jac18281828 replied:
💯💯 thank you this feedback is so helpful, I did this method kind of fast and forgot to even think about From 🙈
